File: bubble.py
#!/usr/bin/env python3
import sys
import os

# A bare bones replication of the dbg! macro in Python
# Just made for easier development


def dbg(elem):
    print(f"DBG ~> {elem}")
    return elem

RED: str = "\033[91m"
GREEN: str = "\033[92m"
RESET: str = "\033[0m"

infobox = {
    "bytemods": f"""{GREEN}Info ~> \n\nBytemods -> Bytemods is a process which intentionally *breaks* a suite of class files for testing{RESET}"""
}
argv: list[str] = sys.argv
WELCOME_MESSAGE: str = """
Welcome to the TapVM Aftermath JVM build system.
The following are the available commands.

# Building.
- buildjvm ~ Builds the JVM alone.
- buildjdk ~ Builds OpenJDK with the built JVM.

# Testing.
- test ~ Tests the JVM.
- advancedtest ~ Advanced testing for the JVM (Requires an internet connection).

# Development utilities.
- classbasket ~ Builds random small Java snippets for JVM testing.
- fmt ~ Formats the source code.
- info [topic] ~ Gives information regarding the JVM about the given topic.

Note -> Set the "BUBBLE_RELEASE" environment variable to "true" to build release builds.
"""

def buildjvm():
    if os.environ.get("BUBBLE_RELEASE") == "true":
        print(f"\n{GREEN}Building production JVM... {RESET}\n")
        if os.system("cargo build --verbose --release") != 0:
            print(f"\n{RED}Build FAILED. :({RESET}\n")
        else:
            print(f"\n{GREEN}Build successful!{RESET}\n")
    else:
        print(
            f"\n{GREEN}Building development JVM... {RED}[DO NOT USE THIS BUILD IN PRODUCTION]{RESET}\n")
        if os.system("cargo build --verbose") != 0:
            print(f"\n{RED}Build FAILED. :({RESET}\n")
        else:
            print(f"\n{GREEN}Build successful!{RESET}\n")


def buildjdk():
    print(f"\n{GREEN}Not so fast cowboy, we ain't there.... yet ;){RESET}\n")


def classbasket():
    print(f"\n{GREEN}Building random small Java snippets for JVM testing...{RESET}")
    if os.system("javac ./class_basket/*.java") != 0:
        print(f"\n{RED}Build FAILED. {RESET}\n")
        exit()
    else:
        print(f"\n{GREEN}Build successful!{RESET}\n")

    os.chdir("scripts/byte_mods/src/")
    print(f"{GREEN}Building byte mods (Run `bubble info bytemods` to know more)\n{RESET}")
    if os.system("cargo run") != 0:
        print(f"\n{RED}Byte mods build FAILED. {RESET}\n")
    else:
        print(f"\n{GREEN}Byte mods build successful!{RESET}\n")


def test():
    classbasket()
    print(f"\n{GREEN}Testing...{RESET}\n")
    if os.system("cargo test") != 0:
        print(f"\n{RED}Tests FAILED. :({RESET}\n")
    else:
        print(f"\n{GREEN}Tests successful!{RESET}\n")


def fmt():
    print(f"\n{GREEN}Formatting...{RESET}")
    if os.system("cargo fmt") != 0:
        print(f"\n{RED}Formatting FAILED. :({RESET}\n")
    else:
        print(f"\n{GREEN}Formatting successful!{RESET}\n")

def info():
    try:
        info: str = argv[2]
    except IndexError:
        print(f"\n{RED}Please provide an info ID. Available options ->{GREEN}")
        for key in infobox.keys():
            print(f"~ {key}")
        print()
        exit(0)

    if infobox.get(info) is not None:
        print(infobox[info])
    else:
        print(f"\n{RED}Please provide a valid info ID. Available options ->{GREEN}")
        for key in infobox.keys():
            print(f"~ {key}")
        print()

def main():
    if len(argv) == 1:
        print(WELCOME_MESSAGE)
        exit()

    command: str = argv[1].lower()

    match command:
        case "buildjvm":
            buildjvm()
        case "buildjdk":
            buildjdk()
        case "test":
            test()
        case "advancedtest":
            advancedtest()
        case "classbasket":
            classbasket()
        case "fmt":
            fmt()
        case "info":
            info()

# --------------------------------------------------------------------------------------------------

main()

File: bubble.jl
#!/usr/bin/env julia

include("./BubbleScripts/namespace.jl")

# ANSI Colors ✨
const RED = "\x1b[31m"
const GREEN = "\x1b[32m"
const BOLD = "\x1b[1m"
const CODE = "\x1b[100m\x1b[30m"
const RESET = "\x1b[0m"

# Messages
const BUILD_FAILURE = "
Note -> If you or any other program didn't edit the source code which might've
caused this build failure, we would greatly appreciate if you made a bug report.
"

function build()
    if length(ARGS) != 1
        if ARGS[2] != "--release"
            println("\n$(RED)Unknown flag$RESET -> $(ARGS[2])\n")
            exit(1)
        else
            try
                run(`cargo run --release`)
            catch e
                println("$(RED)Build failed -> $(e).$RESET")
                println(BUILD_FAILURE)
                exit(1)
            end
        end
    end

    try
        run(`cargo run`)
    catch e
        println("$(RED)Build failed -> $(e).$RESET")
        println(BUILD_FAILURE)
        exit(1)
    end
end

function test()
    try
        run(`cargo test`)
    catch e
        println("$(RED)Build failed -> $(e).$RESET")
        println(BUILD_FAILURE)
        exit(1)
    end
end

function fmt()
    try
        run(`cargo fmt`)
    catch e
        println("$(RED)Build failed -> $(e).$RESET")
        println(BUILD_FAILURE)
        exit(1)
    end
end

function buildjdk()
    println("$GREEN We're not there... yet. $RESET😉")
    exit(0)
end

function advancedtest()
    println("$GREEN We're not there... yet. $RESET😉")
    exit(0)
end

function classbasket()
    print("\n$(GREEN)Building classbasket")

    try
        for file in filter((z) -> endswith(z, ".java"), readdir("./class_basket"))
            run(`javac ./class_basket/$file`)
            print(".")
        end
        println(" Done! $(RESET)✨\n")
        println("Building $(RED)invalid$RESET classfiles...")
        intentionally_invalid()
    catch e
        println("$(RED)Build failed -> $e.$RESET")
        println(BUILD_FAILURE)
        exit(1)
    end

    println("$RESET")
end

if length(ARGS) == 0
    println("""

$GREEN Welcome to the TapVM Aftermath JVM build system. $RESET

$BOLD Commands $RESET
--------------------------------------------------------------------------------
⦁ build        ~ Builds the JVM.
⦁ build-jdk    ~ Builds an OpenJDK distribution using Aftermath.
⦁ test         ~ Tests the JVM.
⦁ advancedtest ~ Advanced testing for the JVM. (Requires an internet connection)
⦁ classbasket  ~ Builds small Java snippets for JVM development and testing.
⦁ fmt          ~ Formats the source code using Rustfmt.

$BOLD Flags $RESET
--------------------------------------------------------------------------------
The $CODE build $RESET and $CODE build-jdk $RESET commands can have an optional $CODE --release $RESET flag. This
will build an optimized and stripped build, however the build time will be more
than a normal build (i.e a$BOLD development$RESET build).
""")
    exit()
else
    command = ARGS[1]
    if command == "build"
        build()
    elseif command == "build-jdk"
        buildjdk()
    elseif command == "test"
        test()
    elseif command == "advancedtest"
        advancedtest()
    elseif command == "classbasket"
        classbasket()
    elseif command == "fmt"
        fmt()
    else
        println("$(RED)Unknown command$RESET -> $command. Please run bubble without any arguments
for help regarding its usage.
")
    end
end
